File: S8-DeepLearning/Pt1-ArtificialNeuralNetworks/artificial_neural_networks.py
## Artificial Neural Networks


## Importing the libraries
import numpy as np 
import matplotlib.pyplot as plt
import pandas as pd


## Part 1 - Data preprocessing

# Importing the dataset
dataset = pd.read_csv('Churn_Modelling.csv')
X = dataset.iloc[:, 3:-1].values
y = dataset.iloc[:, -1].values

# Encoding categorical data
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
X[:, 2] = le.fit_transform(X[:, 2])

# One Hot Encoding the "Geography" column
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
ct = ColumnTransformer(transformers=[('encoder', OneHotEncoder(), [1])], remainder='passthrough')
X = np.array(ct.fit_transform(X))

# Splitting the data into the training set and test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.2, random_state=0)

# Feature Scaling
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)


## Part 2 - Building the ANN
import tensorflow as tf

# Initializing the ANN
ann = tf.keras.models.Sequential()

# Adding the input layer and the first hidden layer 
ann.add(tf.keras.layers.Dense(units=6, activation='relu'))

# Adding the second hidden layer
ann.add(tf.keras.layers.Dense(units=6, activation='relu'))

# Adding the output layer
ann.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))


## Part 3 - Training the ANN

# Compiling the ANN
ann.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

# Training the ANN on the training set
ann.fit(X_train, y_train, batch_size=32, epochs=100)


## Part 4 - Making the predictions and evaluating the model

# Predicting the result of a single observation
'''
Use our ANN model to predict if the customer with the following informations will leave the bank:
Geography: France
Credit Score: 600
Gender: Male
Age: 40 years old
Tenure: 3 years
Balance: $60000
Number of Products: 2
Does this customer have a credit card? Yes
Is this customer an active member? Yes
Estimated Salary: $50000
So, should we say goodbye to that customer?
'''
pred = ann.predict(sc.transform([[1.0, 0.0, 0.0, 600, 1, 40, 3, 60000, 2, 1, 1, 50000]]))
print(pred)
print(pred > .5)

# Predicting the test set results
y_pred = ann.predict(X_test)
y_pred = y_pred > .5
print(np.concatenate((y_test.reshape(len(y_test), 1), y_pred.reshape(len(y_pred), 1)), 1))

# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
print(cm)
print(accuracy_score(y_test, y_pred))

File: S8-DeepLearning/Pt1-ArtificialNeuralNetworks/artificial_neural_networks.R
## Artificial Neural Network

## Importing the dataset
dataset = read.csv('Churn_Modelling.csv')
dataset = dataset[4:14]

## Encoding the categorical variables as factors
dataset$Exited = factor(dataset$Exited, levels = c(0, 1))
dataset$Geography = as.numeric(factor(dataset$Geography, 
                           levels = c('France', 'Spain', 'Germany'), 
                           labels = c(1, 2, 3)))
dataset$Gender = as.numeric(factor(dataset$Gender,
                        levels = c('Male', 'Female'),
                        labels = c(1, 2)))

## Splitting the dataset into the training set and test set
#install.packages('caTools')
library(caTools)
set.seed(123)
split = sample.split(dataset$Exited, SplitRatio = .8)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

## Feature Scaling
training_set[-11] = scale(training_set[-11])
test_set[-11] = scale(test_set[-11])

## Fitting ANN to the training set
#install.packages('h2o')
library(h2o)
h2o.init(nthreads = -1)
classifier = h2o.deeplearning(y = 'Exited',
                              training_frame = as.h2o(training_set),
                              activation = 'Rectifier',
                              hidden = c(6, 6),
                              epochs = 100,
                              train_samples_per_iteration = -2)

## Predicting the test set results
y_pred = h2o.predict(classifier, newdata = as.h2o(test_set[-11]))
y_pred = as.vector(as.numeric(y_pred[, 1]) > .5)

## Making the Confusion Matrix
cm = table(test_set[, 11], y_pred)
cm

## After everything complete
h2o.shutdown()
